iquip/apps/dataviewer.py

```diff
@@ -622,19 +629,122 @@ def __init__(self, parent: Optional[QWidget] = None):
+        self,
+        dataset: np.ndarray,
+        parameters: Sequence[str],
+        units: Sequence[Optional[str]],
+    ):
+        """
+        Args:
+            See attribute section.
+        """
+        self.dataset = dataset
+        self.parameters = parameters
+        self.units = units
+
+    def symbolize(self, axis: Iterable[int]) -> Tuple[List[np.ndarray], np.ndarray]:
+        """Returns the list of unique parameters and symbolized parameter ndarray.
+        
+        It first identifies the unique parameter values in the dataset.
+        These unique values are called "symbols", and hence the process "symbolize".
+        Specifically, symbols are defined by the indices in the unique parameter
+          array (params), i.e., the "inverse" of np.unique().
+        
+        Args:
+            axis: Indices of interested axes. The other axes will be reduced.
+              Note that the index starts from 0, i.e., the index in paremeters,
+              not the dataset column index.
+        
+        Returns:
+            The list of unique parameter values and the symbolized parameter ndarray,
+              where each row corresponds to a parameter, i.e., the shape is
+              (#parameters, #data).
+        """
+        params_list: List[np.ndarray] = []
+        symbols_list: List[np.ndarray] = []
+        for index in axis:
+            params, symbols = np.unique(self.dataset[:, index+1], return_inverse=True)
+            params_list.append(params)
+            symbols_list.append(symbols)
+        return params_list, np.vstack(symbols_list)
+
+    # pylint: disable=too-many-locals
+    def extract(
+        self,
+        axis: Sequence[int],
+        reduce: Callable[[npt.ArrayLike], Any],
+    ) -> Tuple[np.ndarray, List[AxisInfo]]:
+        """Returns the reduced data ndarray and axes information.
+        
+        Args:
+            axis: Indices of interested axes. The other axes will be reduced.
+              Note that the index starts from 0, i.e., the index in paremeters,
+              not the dataset column index.
+            reduce: The function for reducing the not-interested axes.
+        """
+        data = self.dataset[:, 0]  # shape=(N,)
+        if not axis:
+            return np.array(reduce(data)), []
+        params_list, symbol_array = self.symbolize(axis)
+        shape = tuple(map(len, params_list))
+        sorted_indices = np.lexsort(np.flip(symbol_array, axis=0))
+        sorted_symbols = symbol_array.T[sorted_indices]  # shape=(N, M)
+        unique_symbols, unique_indices = np.unique(
+            sorted_symbols, return_index=True, axis=0,
+        )
+        # construct the reduced data array
+        data_groups = np.split(data[sorted_indices], unique_indices[1:])
+        reduced = np.zeros(shape)
+        for index, data_group in zip(unique_symbols, data_groups):
+            reduced[tuple(index)] = reduce(data_group)
+        # sort by the actual parameter
```

Comment:
Is the sorting necessary?
The `symbolize()` have already sorted `params_list` using `np.unique()`, so I think it is unnecessary...

But I may misunderstand it, please notice me if it is necessary😃

Reply:
Oh I didn't know that `np.unique()` sorts the unique values!
What a wonderful review 👍 
I will remove the meaningless sorting part.